### ISSUE TITLE
Error occurred when finalizing GeneratorDataset iterator

### ISSUE THREAD

**Issue by bertelschmitt** [AntonMu/TrainYourOwnYOLO#198]
While putting the last touches on the multi-stream-multi-model-multi-GPU YOLO (out any minute) I notice that training aborts with a  "W tensorflow/core/kernels/data/generator_dataset_op.cc:103] Error occurred when finalizing GeneratorDataset iterator: Failed precondition: Python interpreter state is not initialized. The process may be terminated."  This happens with the factory cat training set on a freshly cloned and untouched TrainYourOwnYOLO.

The model is usable, however it stops at a loss of 15, which is a bit high:

```
Epoch 81/102
22/22 [==============================] - 5s 206ms/step - loss: 15.3284 - val_loss: 13.8459
Epoch 82/102
22/22 [==============================] - 5s 208ms/step - loss: 14.5368 - val_loss: 14.2095
Epoch 83/102
22/22 [==============================] - ETA: 0s - loss: 14.7401
Epoch 00083: ReduceLROnPlateau reducing learning rate to 9.999999939225292e-10.
22/22 [==============================] - 5s 207ms/step - loss: 14.7401 - val_loss: 13.9430
Epoch 84/102
22/22 [==============================] - 5s 208ms/step - loss: 14.1699 - val_loss: 15.3087
Epoch 00084: early stopping
2020-10-31 05:50:12.560589: W tensorflow/core/kernels/data/generator_dataset_op.cc:103] Error occurred when finalizing GeneratorDataset iterator: Failed precondition: Python interpreter state is not initialized. The process may be terminated.

```

I can reproduce this consistently on two separate Ubuntu machines, both with large amounts of memory (64 Gbyte and 256 Gbytes. All libraries stock as per prerequisites. CUDA 10.1, Python3.7, Ubuntu 10. [The problem has received mentions elsewhere](https://www.google.com/search?q=W+tensorflow%2Fcore%2Fkernels%2Fdata%2Fgenerator_dataset_op.cc%3A103%5D+Error+occurred+when+finalizing+GeneratorDataset+iterator%3A+Failed+precondition%3A+Python+interpreter+state+is+not+initialized.&oq=W+tensorflow%2Fcore%2Fkernels%2Fdata%2Fgenerator_dataset_op.cc%3A103%5D+Error+occurred+when+finalizing+GeneratorDataset+iterator%3A+Failed+precondition%3A+Python+interpreter+state+is+not+initialized.&aqs=chrome..69i57j69i64l3.795j0j7&sourceid=chrome&ie=UTF-8)

**Comment by AntonMu** [comment]
Hi @bertelschmitt ,

I'm aware of this error message but I believe everything still works fine despite this message. Did you experience any issues using the trained weights?

**Comment by bertelschmitt** [comment]
Have not looked more into it. If you say that 15 is good for the set, then we can ignore it. I'll  look at it more with my other sets later. The weights work, cat are being recognized.

**Comment by bertelschmitt** [comment]
Multi-stream-multi-model-multi-GPU YOLO will be up tomorrow, documentation cosmetics.  It's already up on Github, but locked. I sent you an early access invite.

**Comment by AndrewW-GIT** [comment]
I've got the same thing  ubuntu 20.04, cuda 10.2 python 3 won't even work with the with the cats

Epoch 00075: early stopping
 W tensorflow/core/kernels/data/generator_dataset_op.cc:103] Error occurred when finalizing GeneratorDataset iterator: Failed precondition: Python interpreter state is not initialized. The process may be terminated.
         [[{{node PyFunc}}]]


Epoch 00088: early stopping
2020-10-31 12:10:48.945618: W tensorflow/core/kernels/data/generator_dataset_op.cc:103] Error occurred when finalizing GeneratorDataset iterator: Failed precondition: Python interpreter state is not initialized. The process may be terminated.
         [[{{node PyFunc}}]]

My CPU is seen but not used, how do i continue ?

**Comment by bertelschmitt** [comment]
@AndrewW-GIT: As discussed above, that error can be ignored. The model should be usable.  If your GPU isn't used, the the issue is elsewhere. Do you see error messages about missing libraries?  If yes, reinstall CUDA. CUDA 10.1 appears to play better with the Tensorflow in prerequisites.  Fire up python. then

```
import tensorflow as tf
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))
```
It should give you the GPUs available to tf. If it gives you and error, reinstall CUDA

**Comment by AndrewW-GIT** [comment]
Thanks Bertslschmitt, I will try your advice about GPUs. I have a GPU detected, and it works with my Yolo code.
when I run this code i get 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties:
pciBusID: 0000:00:08.0 name: GeForce RTX 2070 SUPER computeCapability: 7.5

The code ran again (on a cpu) and exited again but i have some files. I take it the file trained_weights_final.h5 is the one to use ?

When i copy this code over my original yolov3.weights my code crashes with 
cv2.error: OpenCV(4.2.0) /home/mydir/opencv-4.2.0/modules/dnn/src/darknet/darknet_io.cpp:712: error: (-213:The function/feature is not implemented) Transpose the weights (except for convolutional) is not implemented in function 'ReadDarknetFromWeightsStream'

Has the format of the file changed ?

Also i have a coco-names file , do i just need to add the name of the added capability to the bottom ?

Thanks in advance

**Comment by AndrewW-GIT** [comment]
Berslschmitt, I've done your tests on the GPU , the results are below

Originally i was getting
Could not load dynamic library 'libcudnn.so.7'; dlerror: libcudnn.so.7: cannot open shared object file: No such file or directory
I only have a file called libcudnn.so.8 so i synlinked to that file to create the v7 version.
The output is below


2020-10-31 19:34:12.910019: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-10-31 19:34:13.623044: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2020-10-31 19:34:13.639020: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:982] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-10-31 19:34:13.639429: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties:
pciBusID: 0000:00:08.0 name: GeForce RTX 2070 SUPER computeCapability: 7.5
coreClock: 1.77GHz coreCount: 40 deviceMemorySize: 7.79GiB deviceMemoryBandwidth: 417.29GiB/s
2020-10-31 19:34:13.639464: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-10-31 19:34:13.640557: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcublas.so.10
2020-10-31 19:34:13.641790: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcufft.so.10
2020-10-31 19:34:13.641964: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcurand.so.10
2020-10-31 19:34:13.643067: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusolver.so.10
2020-10-31 19:34:13.643662: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusparse.so.10
2020-10-31 19:34:13.647915: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudnn.so.7
2020-10-31 19:34:13.648081: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:982] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-10-31 19:34:13.648729: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:982] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-10-31 19:34:13.649307: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0

**Comment by bertelschmitt** [comment]
Well, it added GPU #0. As I suspected, it didn't find the library. Symlinking  to another one is the easy way out, but is asking for trouble. I suggest a COMPLETE removal of CUDA, and a working install of CUDA 10.1  Follow the Quickie CUDA 101. Works for Ubuntu 10

**Comment by AndrewW-GIT** [comment]
Thanks the reinstall works, although my opencv code always worked

**Comment by bertelschmitt** [comment]
@AndrewW-GIT:  YOLO training does not involve opencv, and opencv works perfectly well without CUDA. You really should familiarize yourself a bit more with the basics.

**Comment by LauChungwei** [comment]
i got the same error message with my 1660(6G), and decreased epoch from default(51) to 35,  then the  error message disappeared. i don't know why by it does works in my case.

**Comment by bertelschmitt** [comment]
@LauChungwei, I'd recommend ignoring the error, and bringing the epochs to 51, or higher. Precision likely will suffer if epochs are set too low.

**Comment by LauChungwei** [comment]
> @LauChungwei, I'd recommend ignoring the error, and bringing the epochs to 51, or higher. Precision likely will suffer if epochs are set too low.

i definitely agree with you on that, and i decreased epoch to 35 cause  loss didn't go down any more in my case . i trained about 400 pictures, and loss keeps wandering around 12.00000 after 60th epoch,  it is no difference  against the final loss after 74th epoch, where i used to be stopped by that error.

**Comment by bertelschmitt** [comment]
If it works, it works. However, you might see that it stops first, and only then throws the meaningless error.

**Comment by Algabri** [comment]
> import tensorflow as tf
> print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))

it is good and worked for me. Also, those lines gave the same result.

```
physical_devices = tf.config.experimental.list_physical_devices('GPU')
    # physical_devices = tf.config.experimental.list_physical_devices('CPU')
    print("physical_devices-------------", len(physical_devices))
    tf.config.experimental.set_memory_growth(physical_devices[0], True)
```
